### ISSUE TITLE
Indent continue not working for class declarations in JAVA

### ISSUE THREAD

**Issue by kaiser101** [uncrustify/uncrustify#1109]
Uncrustify version - Uncrustify-0.64-569-a6cd6af

Expected result
For the class declaration it should be indented by 8 spaces on the next line.
```
public abstract class KeyValueItemWriter<K, V> implements ItemWriter<V>,
        InitializingBean {
```

Actual result
The class declaration is indented by only 4 spaces.
```
public abstract class KeyValueItemWriter<K, V> implements ItemWriter<V>,
    InitializingBean {
```
[debug.txt](https://github.com/uncrustify/uncrustify/files/911130/debug.txt)

**Comment by kaiser101** [comment]
Attaching the file being uncrustified.


[KeyValueItemWriter.docx](https://github.com/uncrustify/uncrustify/files/911151/KeyValueItemWriter.docx)

**Comment by CDanU** [comment]
I am not going to open that docx file. Please upload your file to gist.github.com.

**Comment by CDanU** [comment]
@gmaurel Could you remove his docx comment?

**Comment by kaiser101** [comment]
@CDanU Yes it is a duplicate of #670 . I have removed the attachment. The reason I attached it as a docx was because github would not allow me to upload the raw text format or even a zip file. I think you can close either of the two issues.

**Comment by kaiser101** [comment]
I will try to upload the attachment again later today.

**Comment by kaiser101** [comment]
[KeyValueItemWriter.java.txt](https://github.com/uncrustify/uncrustify/files/914585/KeyValueItemWriter.java.txt)

**Comment by CDanU** [comment]
I debugged this a bit, in order to indent the line the code must go trough this steps (this is one way which works, most likely there is a better approach):

1) [indent.cpp#L2017-L2027](https://github.com/uncrustify/uncrustify/blob/a6cd6afc7fb82f07d91fb2d55069aa02e0423454/src/indent.cpp#L2017-L2027)

When the pc chunk is storing the info for 'InitializingBean' `vardefcol` needs to be `0` and `pc->flags` must be `PCF_VAR_1ST_DEF` in order to call `calc_indent_continue()`

The problem here is that the `pc` of 'InitializingBean' does not have the right flag set and the value of `vardefcol` is previously set when `pc` is 'ItemWriter'.

2) [indent.cpp#L2055-L2096](https://github.com/uncrustify/uncrustify/blob/a6cd6afc7fb82f07d91fb2d55069aa02e0423454/src/indent.cpp#L2055-L2096)

`do_vardefcol` needs to be set to `true` in line 2094 to call `reindent_line(pc, vardefcol);` in line 2125.
This does not happen because of the flag problem (2091).

**Comment by kaiser101** [comment]
Can someone fix this please?

**Comment by guy-maurel** [comment]
PR #1308 doesn't apply here.

**Comment by kaiser101** [comment]
Can someone please help me out with this issue?

**Comment by nevercast** [comment]
I'm not savvy enough to provide a patch for this unfortunately, but something that I'd consider a temporary solution is using Google's Java formatter. Obviously because it's Google's there is no option to change the style, it just follows their in-house standard. But for me it will tide me over until I can use Uncrustify again.

**Comment by kaiser101** [comment]
@nevercast I tried using google java formatter, but the problem with it is that it uses 2 spaces for indenting. The above issue is the only thing holding me back from using uncrustify, because I have to format long class names manually again. For now I have fallen back to using Eclipse formatting via a sublime plugin.

**Comment by kaiser101** [comment]
Can someone please help me out with this issue?

**Comment by guy-maurel** [comment]
The current git produces:
```
public abstract class KeyValueItemWriter <K, V> implements ItemWriter <V>, InitializingBean {
    protected Converter <V, K> itemKeyMapper;
    protected boolean delete;
    }
```
What do you think about it?

**Comment by CDanU** [comment]
#670 has a better description of the problem. In my previous ramblings I tried to force the indent like a vardef, this is not a good idea. I guess this will actually need new code to handle the continuation indent of the `implements` token.

**Comment by kaiser101** [comment]
@guy-maurel The problem arises when the line is wrapped and moved to the next line. It should be indented by 2 indents instead of one to separate it from the variable/constructor/method indents. This helps in avoiding confusion.

@CDanU Actually #670 was something which I had raised. I could not get any traction going so I created a new one for the class definition indent. I am not sure if we need new code, The reason is because the statement could wrap at tokens other than implements or extends. Basically what it should do is it should apply the default indent to a class definition statement, which I believe it currently does not, because the class definition is at the top of the hierarchy. If it is able to apply a single indent for the long class definition statement when it gets wrapped, then combined with indent continue, we will get the expected behavior. I hope I am making sense.

**Comment by kaiser101** [comment]
Can someone look into this please?

**Comment by guy-maurel** [comment]
Which line do you mean?
What is wrong at the output I proposed on Dec 15, 2017?
Could you modify it to what you expected to get?

**Comment by kaiser101** [comment]
Hi Guy,

The problem with that solution is when the class declaration gets split via the code_width flag (code_width = 80). The next line is indented by one indent width (by 4 spaces if indent_columns = 4). This should not be the case. Java convention mandates `that` in such cases the line should be indented by 2 indents to help distinguish it from the statements which follow, for e.g. variable declarations, constructors, methods, etc.

```
public abstract class KeyValueItemWriter<K, V> implements ItemWriter<V>,
        InitializingBean {
    protected Converter<V, K> itemKeyMapper;
    protected boolean delete;
}
```

The above output is from Eclipse/NetBeans standard Java formatting.

**Comment by guy-maurel** [comment]
In a first pass, after a new line is created, the indent of` InitializingBean` is set to 9.
At the next pass, `indent_text` changes it back to 5.
I don't know how to prevent that.

**Comment by guy-maurel** [comment]
A change such as:
```
diff --git a/src/width.cpp b/src/width.cpp
index 487259d..40aaecb 100644
--- a/src/width.cpp
+++ b/src/width.cpp
@@ -143,6 +143,14 @@ static void split_before_chunk(chunk_t *pc)
       // reindent needs to include the indent_continue value and was off by one
       reindent_line(pc, pc->brace_level * cpd.settings[UO_indent_columns].u +
                     abs(cpd.settings[UO_indent_continue].n) + 1);
+      // Issue #1109
+      if (  (cpd.lang_flags & LANG_JAVA)
+         && !(pc->flags & (PCF_IN_FOR)))
+      {
+         LOG_FMT(LSPLIT, "%s(%d): text() '%s', flags: \n", __func__, __LINE__, pc->text());
+         log_pcf_flags(LGUY, pc->flags);
+         pc->flags |= PCF_DONT_INDENT;
+      }
       cpd.changes++;
    }
 }
```
is not stable. The second run destroy the indentation.

**Comment by kaiser101** [comment]
The problem is specific only to a class definition. Variable and method declarations get indented just fine. Something special is happening with a class definition in Java, causing the second run to rest the indent level to 5.

**Comment by gmaurel** [comment]
What special?
How can uncrustify recognize it?
We need anything to set a flag on the token?

**Comment by kaiser101** [comment]
I am not very well versed with the code, nor am I very comfortable in c++. I think we do have a flag, but I think we need to use it to stop the second run from somehow resetting the indent level to 5. Can we somehow use the CT_CLASS along with LANG_JAVA to identify that it is a class definition and prevent the reset?

**Comment by CDanU** [comment]
I, E:
```cpp
int main()
{
    int i = 20000;
    while(i--);
    return 0;
}
```

O:
```cpp
int main()
{
    int i = 20000;
    while (i--)
        ;
    return 0;
}
```
C:
```sql
# Add or remove newline between 'while' and '{'.
nl_while_brace                  = force    # ignore/add/remove/force

# Change a one-liner while statement into simple unbraced while
# 'while (i<5) foo(i++);' => 'while (i<5)\n foo(i++);'.
nl_split_while_one_liner        = true     # false/true

```

**Comment by kaiser101** [comment]
Does the pull request give us any clue about the problem? Why does the second pass reset the indent level to 5?

**Comment by guy-maurel** [comment]
Duplicate to #672

**Comment by guy-maurel** [comment]
PR #2204 is a bugfix proposal.
Please test it and report.
